idaes/core/initialization/initializer_base.py

```diff
@@ -0,0 +1,375 @@
+        if not degrees_of_freedom(model) == 0:
+            self.status = InitializationStatus.DoF
+            raise InitializationError(
+                f"Degrees of freedom for {model.name} were not equal to zero during "
+                f"initialization (DoF = {degrees_of_freedom(model)})."
+            )
+
+    def initialization_routine(self, model: Block):
+        """
+        Placeholder method to run initialization routine. Derived classes should overload
+        this with the desired routine.
+
+        Args:
+            model: Pyomo Block to initialize.
+
+        Returns:
+            Overloaded method should return a Pyomo solver results object is available,
+            otherwise None
+
+        Raises:
+            NotImplementedError
+        """
+        self.status = InitializationStatus.Error
+        raise NotImplementedError()
+
+    def restore_model_state(self, model: Block):
+        """
+        Restore model state to that stored in self.initial_state.
+
+        Args:
+            model: Pyomo Block ot restore state on.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError if no initial state is stored.
+        """
+        if self.initial_state is not None:
+            from_json(model, sd=self.initial_state, wts=StoreState)
+        else:
+            self.status = InitializationStatus.Error
+            raise ValueError("No initial state stored.")
+
+    def postcheck(self, model: Block, results_obj: dict = None):
+        """
+        Check the model has been converged after initialization.
+
+        If a results_obj is provided, this will be checked using check_optimal_termination,
+        otherwise this will walk all constraints in the model and check that they are within
+        tolerance (set via the Initializer constraint_tolerance config argument).
+
+        Args:
+            model: model to be checked for convergence.
+            results_obj: Pyomo solver results dict (if applicable, default=None).
+
+        Returns:
+            InitialationStatus Enum
+        """
+        if results_obj is not None:
+            self.postcheck_summary = {
+                "solver_status": check_optimal_termination(results_obj)
+            }
+            if not self.postcheck_summary["solver_status"]:
+                # Final solver call did not return optimal
+                self.status = InitializationStatus.Failed
+                raise InitializationError(
+                    f"{model.name} failed to initialize successfully: solver did not return "
+                    "optimal termination. Please check the output logs for more information."
+                )
+        else:
+            # Need to manually check initialization
+            # First, check that all non-stale Vars have values
+            uninit_vars = []
+            for v in model.component_data_objects(Var, descend_into=True):
+                if v.value is None:
+                    uninit_vars.append(v)
```

Comment:
At the moment it does not. I originally intended it to, but the `stale` attribute on components does not get set until a solver has been called so I resorted to just checking all `Vars` until I could come up with an alternative (preferably one that does not involve walking all the expressions). Suggestions on how to do this would be welcome.

Reply:
I would default to just walking the expressions. If this is done before for this form of the model, it could be cached somewhere? I don't see an easy way to do this with `stale`, partially because I'm not 100% clear on how stale behaves and partially because we have many different solves (and types of solves, i.e. pure python calculation vs external solver) that could happen in the initialization routine.

Reply:
But I agree that doing this without walking the expressions would be desirable if possible.